sycl/include/sycl/detail/accessor_iterator.hpp

```diff
@@ -0,0 +1,375 @@
+  // incorporating that offset into _MLinearId right away.
+  //
+  // _MPerRowOffset stores a number of _inaccessible_ elements in each
+  // _accessible_ row. For the example above it would be equal to 2 (leftmost
+  // and the rightmost elements of a row).
+  //
+  // _MPerSliceOffset stores a number of _inaccessible_ elements in each
+  // _accessible_ slice. Slice here means a single 2D layer in a 3D buffer. For
+  // the example above it would be equal to 0, because we are not looking at a
+  // 3D buffer. However, if we had two slices like visualized above,
+  // _MPerSliceOffset would be equal to 16 (elements on the "perimeter" of the
+  // slice, i.e. ones represented as dots (.)).
+
+  size_t _MStaticOffset = 0;
+  size_t _MPerRowOffset = 0;
+  size_t _MPerSliceOffset = 0;
+
+  // Contains a number of _accessible_ elements in a row
+  size_t _MRowSize = 0;
+  // Contains a number of _accessible_ elements in a slice
+  size_t _MSliceSize = 0;
+
+  // _MLinearId stores an offset which is relative to the accessible range of
+  // the accessor, which means that it could be the case that _MlinearId equal
+  // to 0 should not correspond to the beginning of the underlying buffer, but
+  // instead should be re-adjusted to account for an offset passed to the
+  // accessor constructor.
+  //
+  // This function performs necessary calculations to make sure that all
+  // access ranges and offsets are taken into account.
+  size_t __get_absolute_offset_to_buffer() {
+    // For 1D case, any possible offsets are already incorporated into
+    // _MLinearId, so 1D is always treated as a non-ranged accessor
+    if (!_MAccessorIsRanged || _Dimensions == 1)
+      return _MLinearId;
+
+    // Here we need to deal with 2D or 3D ranged accessor.
+    // _MLinearId points to an element relative to the accessible range. It
+    // should be adjusted to account for elements which are outside of the
+    // accessible range of the accessor.
+
+    // We start with static offset: that is a number of elements in full rows
+    // and full slices before the first accessible element.
+    size_t _AbsoluteId = _MLinearId + _MStaticOffset;
+
+    // Then we account for inaccessible elements in each full slice
+    size_t _Remaining = _MLinearId;
+    if constexpr (_Dimensions == 3) {
+      _AbsoluteId += _MPerSliceOffset * (_Remaining / _MSliceSize);
+      _Remaining %= _MSliceSize;
+    }
+
+    // Then we account for inaccessible elements in each full row
+    _AbsoluteId += _MPerRowOffset * (_Remaining / _MRowSize);
+    _Remaining %= _MRowSize;
+
+    return _AbsoluteId;
+  }
+
+  __accessor_iterator(_DataT *_DataPtr, const range<_Dimensions> &_MemoryRange,
+                      const range<_Dimensions> &_AccessRange,
+                      const id<_Dimensions> &_Offset)
+      : _MDataPtr(_DataPtr) {
+    constexpr int _XIndex = _Dimensions - 1;
+    constexpr int _YIndex = _Dimensions - 2;
+    (void)_YIndex;
```

Comment:
Very much a nit, but I would prefer we lean into using `std::ignore` for these situations.

Reply:
So, looking at [cppreference](https://en.cppreference.com/w/cpp/utility/tuple/ignore) I don't quite understand how exactly should I use it in here. 

From what I understand `std::ignore` simply defines a custom `operator=`, which allows to silence the warning, but those values here are not always ignored

Reply:
Should just be something like 
```suggestion
    std::ignore = _YIndex;
```